lib/vm-agent.js

```diff
@@ -9,450 +9,696 @@
- * passed it means that a vm is tentatively destroyed and we double check that
- * there. In case the vm is not destroyed we reload its data with vmadm lookup.
- * At the end we call updateVm to update the vm data on VMAPI
+ * This is called when a VM has failed an update. It is responsible for waiting
+ * for (and incrementing) the delay and then re-queuing the update if there's
+ * not already a timer doing the same thing.
  */
-VmAgent.prototype.sendSample = function (options, callback) {
+VmAgent.prototype.retryFailedUpdate = // eslint-disable-line
+function retryFailedUpdate(err, vm_uuid, callback) {
     var self = this;
-    var log = this.log;
-    var vmapiClient = this.vmapiClient;
-
-    // if vm was destroyed else vmadm lookup
-    if (options.cachedVm && options.cachedVm.state === 'destroyed') {
-        var destSample = {};
-        destSample[options.uuid] = options.cachedVm;
-        this.setSample(destSample);
-        this.eventWatcher.removeState(options.uuid);
-
-        vmapiClient.updateVm(options.cachedVm, callback);
-
-    } else {
-        this.updateSample(options, function (err, sample) {
-            if (err) {
-                log.error(err, 'updateSample failed, cannot sendSample');
-                callback(err);
-
-            } else if (Object.keys(sample).length === 0) {
-                log.warn('empty sample returned by vmadm lookup');
-                callback();
-
-            } else {
-                self.eventWatcher.updateState(sample);
-                self.setSample(sample);
-                var vm = sample[options.uuid];
-                vmapiClient.updateVm(vm, callback);
-            }
-        });
-    }
-};
+    var delay;
 
+    if (self.retryDelays[vm_uuid] && self.retryDelays[vm_uuid].timer) {
+        // Had an error, but we also already have a timer, so don't start
+        // another one, or increment the delay.
+        callback();
+        return;
+    }
 
-/*
- * Initializes the serial queue for a single uuid and pushes the item to be
- * processed.
- */
-VmAgent.prototype.pushSample = function (options) {
-    var uuid = options.uuid || this.uuid;
+    // On any error when we don't already have a timer running, we set a
+    // timer to re-queue the VM. This is always safe since in the worst
+    // case we do an extra update with the latest data.
+    if (!self.retryDelays[vm_uuid]) {
+        self.retryDelays[vm_uuid] = {delay: INITIAL_UPDATE_DELAY};
+    }
+    delay = self.retryDelays[vm_uuid].delay;
 
-    if (this.queues[uuid] === undefined) {
-        this.queues[uuid] = this.createQueue(uuid);
+    // Increment for next time.
+    self.retryDelays[vm_uuid].delay *= 2;
+    if (self.retryDelays[vm_uuid].delay > MAX_UPDATE_DELAY) {
+        self.retryDelays[vm_uuid].delay = MAX_UPDATE_DELAY;
     }
 
-    this.queues[uuid].push(options);
-};
+    self.log.warn({err: err, vm_uuid: vm_uuid}, 'update failed, will try '
```

Comment:
I would move that logging call in the caller, because that's really where we know that the update failed.